bin/test-cleanup: Delete test namespaces in paralllel.

briansmith opened the pull request:
My kubectl (1.11.1) waits until the `kubectl delete` operation completes,
unlike previous versions which issue the deletion request and then
immediate return. As a result of this change, bin/test-cleanup become
much slower than it used to be in my configuration. Fix this by issuing
all the deletions in parallel with a single `kubectl delete`.

Also, make bin/test-cleanup exit with a failure status code if no
namespaces are deleted. Previously, with kubectl 1.11.1, the script
would exit with a success error code when no namespaces were
deleted.

Signed-off-by: Brian Smith <brian@briansmith.org>

briansmith commented on bin/test-cleanup:
```diff
@@ -9,7 +9,4 @@ fi
-for ns in $(kubectl get ns | grep "^$linkerd_namespace-" | cut -f1 -d ' '); do
-  kubectl delete ns "$ns"
-done
+kubectl get ns | grep "^$linkerd_namespace-" | cut -f1 -d ' ' | xargs kubectl delete ns "$linkerd_namespace"
```
> The script as written will return an error if I try to cleanup a namespace that doesn't exist, which then prompts me to cleanup the correct namespace.

I verified that the updated script will exit with a failure exit code (and prints a message) if there is no `linkerd` namespace and there are no `linkerd-*` namespaces. I think this is the right behavior; if there are `linkerd-*` namespaces then they should be deleted even if there is no `linkerd` namespace.

briansmith replied:
> Also, I'm running kubectl v1.9.7 and I don't see the delay issue that you're seeing. Maybe it's a regression?

IDK. It might be a configuration setting on my part. However, the waiting behavior is very useful so I'd like to accomodate it.

klingerf commented on bin/test-cleanup:
```diff
@@ -9,7 +11,6 @@ fi
-done
+# `kubectl delete ns` will fail if there are no matching matching namespaces,
+# causing this script to exit with an error status.
+kubectl get ns | cut -f1 -d ' ' | grep -E "^$linkerd_namespace(-|$)" | xargs kubectl delete ns
```
Thanks for updating! This still doesn't return an error for me if there are no matching namespaces:

```
$ ./bin/test-cleanup asdf; echo $?
cleaning up namespace [asdf] and associated test namespaces
0
```

I think the issue is that my version of kubectl behaves like this:

```
$ echo "" | xargs kubectl delete ns; echo $?
0
```

briansmith commented:
> kubectl v1.9.7

I believe the reason you're seeing different behavior than me is because you are using an old version of kubectl. I am using v1.11.1:

````sh
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.1", GitCommit:"b1b29978270dc22fecc592ac55d903350454310a", GitTreeState:"clean", BuildDate:"2018-07-17T18:53:20Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.4", GitCommit:"bee2d1505c4fe820744d26d41ecd3fdd4a3d6546", GitTreeState:"clean", BuildDate:"2018-03-12T16:21:35Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
```

The default behavior of `kubectl delete` changed in https://github.com/kubernetes/kubernetes/pull/64034.

So, I suggest we solve this by keeping the PR as-is and adding a note to TEST.md that one should use kubectl 1.11.1 or later.

briansmith commented:
I added the note to TEST.md.

I also tested the behavior of `bin/test-cleanup` on the master branch (without this PR) and found that it actually is broken w.r.t. the exit code:

```sh
$ bin/test-cleanup asdfasdf && echo "it worked"
cleaning up namespace [asdfasdf] and associated test namespaces
Error from server (NotFound): namespaces "asdfasdf" not found
it worked
```

So this PR actually fixes a bug, as well. I've updated the PR description accordingly.

klingerf commented on TEST.md:
```diff
@@ -84,6 +84,7 @@ cluster. Prior to running the test suite, verify that:
 
 - The Conduit docker images you're trying to test have been built and are
   accessible to the Kubernetes cluster to which you are deploying
+- `kubectl version` reports the client version as v1.11.1 or later.
```
I use gcloud to manage my kubectl installation, and gcloud's version of kubectl is at v1.9.7 currently ([release notes](https://cloud.google.com/sdk/docs/release-notes#kubernetes_engine_4)). I have a strong preference for our minimum required version to not exceed the version that's available via gcloud.

briansmith replied:
> I have a strong preference for our minimum required version to not exceed the version that's available via gcloud.

What's the rationale?

briansmith replied:
> What's the rationale?

Never mind.

I just installed kubectl through gcloud and the version is "v1.11.0."

klingerf replied:
My rationale is that we shouldn't require the latest available version of kubectl to delete a few namespaces; I would rather update the script so that it works regardless of kubectl version.

I think something like this would work:

```
if ! namespaces=$(kubectl get ns -oname | grep "/$linkerd_namespace"); then
  echo "no namespaces found for [$linkerd_namespace]"
  exit 1
else
  kubectl delete $namespaces
fi
```

That will exit with an error if no namespaces exist, and otherwise it will delete all namespaces in parallel.

briansmith commented:
I made the changes you suggested.
